crates/bevy_sprite/src/render/mod.rs

```diff
@@ -620,154 +631,162 @@ pub fn prepare_sprites(
+                    * Affine3A::from_scale_rotation_translation(
+                        quad_size.extend(1.0),
+                        Quat::IDENTITY,
+                        (quad_size * (-extracted_sprite.anchor - Vec2::splat(0.5))).extend(0.0),
+                    );
+
+                // Store the vertex data and add the item to the render phase
+                sprite_meta
+                    .sprite_instance_buffer
+                    .push(SpriteInstance::from(
+                        &transform,
+                        &extracted_sprite.color,
+                        &uv_offset_scale,
+                    ));
+                if batch_image_changed {
+                    batch_item_index = item_index;
+
+                    let new_batch = SpriteBatch {
+                        image_handle_id: batch_image_handle,
+                        range: index..index,
+                    };
+
+                    batches.push((item.entity, new_batch.clone()));
                 }
+
+                transparent_phase.items[batch_item_index].batch_size += 1;
+                batches.last_mut().unwrap().1.range.end += 1;
+                index += 1;
             }
         }
         sprite_meta
-            .vertices
-            .write_buffer(&render_device, &render_queue);
-        sprite_meta
-            .colored_vertices
+            .sprite_instance_buffer
             .write_buffer(&render_device, &render_queue);
+
+        let total_indices = 6;
+        let current_indices = sprite_meta.sprite_index_buffer.len();
+        if total_indices > current_indices {
+            sprite_meta
+                .sprite_index_buffer
+                .values_mut()
+                .reserve(total_indices - current_indices);
+            for i in (current_indices / 6)..total_sprites {
+                let base = (i * 4) as u32;
+                // NOTE: This code is generating sets of 6 indices pointing to 4 vertices.
+                // The vertices form the corners of a quad based on their two least significant bits.
+                // 01   11
+                //
+                // 00   01
+                // The sprite shader can then use the two least significant bits as the vertex index,
+                // and the remaining upper bits as the instance index. The rest of the properties to
+                // transform the vertex positions and UVs (which are implicit) are baked into the
+                // instance transform, and UV offset and scale. See bevy_sprite/src/render/sprite.wgsl
+                // for the details.
+                sprite_meta.sprite_index_buffer.push(base + 2);
+                sprite_meta.sprite_index_buffer.push(base);
+                sprite_meta.sprite_index_buffer.push(base + 1);
+                sprite_meta.sprite_index_buffer.push(base + 1);
+                sprite_meta.sprite_index_buffer.push(base + 3);
+                sprite_meta.sprite_index_buffer.push(base + 2);
```

Comment:
I don't have much context of how the sprite rendering works, but is the index buffer needed if it just contains the same offsets over and over again?

i.e. would it be possible to base the indices off of `@builtin(vertex_index)` inside the vertex shader and read the coordinates from a buffer?  (or generate the index buffer on the GPU if one is strictly needed)

Reply:
There is something called a 'post-transform cache' where indexed vertices that use the same index have their vertex shader results cached and reused. I think if you used unique built-in indices just specifying a range to the draw command, then you would invoke the vertex shader 6 times per quad instead of 4.

Reply:
Good point, thanks for elaborating. I had something along the lines of vertex pulling in mind, but I'm fairly certain the caching benefits are lost in that case as you mentioned.